components/core/src/templating/mod.rs

```
@@ -0,0 +1,414 @@
+// Copyright (c) 2016-2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod config;
+mod context;
+pub mod health;
+pub mod helpers;
+pub mod hooks;
+pub mod package;
+#[cfg(test)]
+pub mod test_helpers;
+
+use std::fmt;
+use std::ops::{Deref, DerefMut};
+use std::result;
+
+use handlebars::{Handlebars, RenderError};
+use serde::Serialize;
+use serde_json;
+
+use error::{Error, Result};
+use fs;
+use package::PackageInstall;
+
+pub use self::context::RenderContext;
+
+pub fn compile_from_package_install(package: &PackageInstall) -> Result<()> {
+    let pkg = package::Pkg::from_install(package.clone())?;
+
+    fs::SvcDir::new(&pkg.name, &pkg.svc_user, &pkg.svc_group).create()?;
+
+    let cfg = config::Cfg::new(&pkg, None)?;
+    let ctx = RenderContext::new(&pkg, &cfg);
+    let cfg_renderer = config::CfgRenderer::new(pkg.path.join("config"))?;
+    cfg_renderer.compile(&pkg.name, &pkg, &ctx)?;
+
+    hooks::HookTable::from_package_install(package).compile(&pkg.name, &ctx);
```

What happens if you try to compile a hook that depends on things that aren't present in this limited rendering context (like census information)?

Seems like we should throw an error.

---

I'm assuming this comment is also targetted to config templates also? Currently it just renders nothing if there is nothing in the context. Thinking of the following example scenario, I'd prefer to raise a warning instead of an error:

There may be multiple config templates and some have, for example, binds and others do not. When the `install` hook fires at install time, it will need to compile all templates since it cannot assume which templates it will use. We wouldn't want the templates with the binds to error during installation. 

Also for debug purposes, one might use `hab pkg compile` and there would be no `bind` context. So a warning informing the cli user that there is missing context would be nice but they could still render the templates.

---

One extra point. I'm not certain yet how much we can control for that via the serde API.

---

I also thought we were going to have a specific install hook, or am I misremembering?

That might end up being cleaner, since we wouldn't necessarily need to bring over all the hooks, just the handlebars and context stuff.

---

Yeah there is an `InstallHook` added to `hooks.rs`. Maybe it makes better sense for the `HookTable` not to include that hook and then just keep `HookTable` and the non-install hook implementations with the supervisor.

---

The downside of keeping the non-`install` hooks with the supervisor is that the `hab pkg compile` cli command will not be able to compile them. Also, future work is going to need to be able to compile the health-check via the cli.